core/src/main/java/org/jdbi/v3/core/statement/SqlStatements.java

```diff
@@ -101,7 +104,9 @@ public Object getAttribute(String key) {
      * @return the defined attributes.
      */
     public Map<String, Object> getAttributes() {
-        return attributes;
+        synchronized (this) {
+            return new HashMap<>(attributes);
```

Comment:
This will break anybody who calls e.g. `getAttributes().put(key, value)`

Reply:
Yeah, it's true, although that's a super suspect use.  Returning a mutable map essentially requires us to continue using `ConcurrentHashMap` (synchronized map isn't good enough -- you need to know that e.g. you have to synchronize iteration) which forbids null values.

We could fix this with a null guard value.  That likely increases the complexity a fair bit.  Do you think it's worth it?  We never promised anything about the mutability or updateability of the returned collection, and as far as I know only JAXB is crazy enough to assume it.

Reply:
tbh I'm with steven on that. It's best to not assume you can just `get()` a collection and then `add()` stuff to it with any useful result.

Reply:
Ok. I would suggest returning `Collections.unmodifiableMap()` in that case, to cement that assumption.